Fix sfp test's starting index

prgeor opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR

The current platform API tests for SFP uses indices starting from 0 which fails to return the correct SFP object for platforms having physical port number starting from 1


Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [x] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)


### Back port request
- [x] 202012

### Approach
#### What is the motivation for this PR?
Fix issues #8435, #8436 buildimage repo

#### How did you do it?
Obtain the physical port indices from config DB. This should work for breakout ports and multi-asic as well.

#### How did you verify/test it?
Ran all tests under test_sfp.py on Arista 7050CX3

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation 
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

prgeor commented:
@rawal01 can you review?

rawal01 commented:
@prgeor 
The change works in almost all cases but I see a case where if you do not have all interfaces used on your DUT configured, in that case `compare_value_with_platform_facts` will not return right values and test using this will fail (right now only test_get_name).

              def compare_value_with_platform_facts(self, key, value, sfp_idx, duthost):
                   expected_value = None
                   sfp_id = self.list_sfps.index(sfp_idx)
                   if duthost.facts.get("chassis"):
                       expected_sfps = duthost.facts.get("chassis").get("sfps")
                   if expected_sfps:
                       expected_value = expected_sfps[sfp_id].get(key)

this because your list_sfps will be just a subset of all interfaces while platform.json from which it is comparing always has all interface sfp names.
One more suggestion  `physical_port_indices' can be moved to platform utils and can be used on other tests like test_sfps in platform_tests/api/test_chassis.py

prgeor commented:
> @prgeor
> The change works in almost all cases but I see a case where if you do not have all interfaces used on your DUT configured, in that case `compare_value_with_platform_facts` will not return right values and test using this will fail (right now only test_get_name).
> 
> ```
>           def compare_value_with_platform_facts(self, key, value, sfp_idx, duthost):
>                expected_value = None
>                sfp_id = self.list_sfps.index(sfp_idx)
>                if duthost.facts.get("chassis"):
>                    expected_sfps = duthost.facts.get("chassis").get("sfps")
>                if expected_sfps:
>                    expected_value = expected_sfps[sfp_id].get(key)
> ```
> 
> this because your list_sfps will be just a subset of all interfaces while platform.json from which it is comparing always has all interface sfp names.
> One more suggestion `physical_port_indices' can be moved to platform utils and can be used on other tests like test_sfps in platform_tests/api/test_chassis.py



> @prgeor
> The change works in almost all cases but I see a case where if you do not have all interfaces used on your DUT configured, in that case `compare_value_with_platform_facts` will not return right values and test using this will fail (right now only test_get_name).
> 
> ```
>           def compare_value_with_platform_facts(self, key, value, sfp_idx, duthost):
>                expected_value = None
>                sfp_id = self.list_sfps.index(sfp_idx)
>                if duthost.facts.get("chassis"):
>                    expected_sfps = duthost.facts.get("chassis").get("sfps")
>                if expected_sfps:
>                    expected_value = expected_sfps[sfp_id].get(key)
> ```
> 
> this because your list_sfps will be just a subset of all interfaces while platform.json from which it is comparing always has all interface sfp names.
> One more suggestion `physical_port_indices' can be moved to platform utils and can be used on other tests like test_sfps in platform_tests/api/test_chassis.py

@rawal01 I have addressed your comment. can you check and approve?

sujinmkang commented:
@prgeor can you fix the lgtm issue while you're waiting for @rawal01 's review?

rawal01 commented:
@prgeor    Thank you for changes it resolves the one case I mentioned and also taking care of test_chassis.

 I would like suggest some changes  `def get_physical_port_indices(duthosts, enum_rand_one_per_hwsku_hostname, conn_graph_facts):` be changed to `def get_physical_port_indices(duthost)' as duthost is already defined in from where we make a call to this fixture.  Also conn_graph_facts is not longer used.

also in both test_chassis.py and test_sfp.py we do not to define physical_pot_indices any more seems uneccessary
        @pytest.fixture(scope="module")
        def physical_port_indices(duthosts, enum_rand_one_per_hwsku_hostname, conn_graph_facts):
              return get_physical_port_indices(duthosts, enum_rand_one_per_hwsku_hostname, conn_graph_facts)
              
I think it is simpler to assign lit directly ` list_sfps = get_physical_port_indices(duthost)`  as fixture is returning a list we need.

prgeor commented:
> @prgeor Thank you for changes it resolves the one case I mentioned and also taking care of test_chassis.
> 
> I would like suggest some changes `def get_physical_port_indices(duthosts, enum_rand_one_per_hwsku_hostname, conn_graph_facts):` be changed to `def get_physical_port_indices(duthost)' as duthost is already defined in from where we make a call to this fixture. Also conn_graph_facts is not longer used.
> 
> also in both test_chassis.py and test_sfp.py we do not to define physical_pot_indices any more seems uneccessary
> @pytest.fixture(scope="module")
> def physical_port_indices(duthosts, enum_rand_one_per_hwsku_hostname, conn_graph_facts):
> return get_physical_port_indices(duthosts, enum_rand_one_per_hwsku_hostname, conn_graph_facts)
> 
> I think it is simpler to assign lit directly ` list_sfps = get_physical_port_indices(duthost)` as fixture is returning a list we need.

@rawal01  It will be easier if you can comment against the file/line number. I have made the changes and fixed LGTM warnings. get_physical_port_indices() is a time consuming function and hence i have wrapped around it in a pytest fixture with module scope.

prgeor commented:
@yxieca could you review

rawal01 commented:
@prgeor Yes makes sense the changes look good to me

yxieca commented on tests/common/platform/interface_utils.py:
```diff
@@ -121,3 +124,26 @@ def get_port_map(dut, asic_index=None):
+	asichost = duthost.asic_instance(asic_index)
+	for intf in interfaces_per_asic:
+	    cmd = 'redis-cli -n 4 hget "PORT|{}" index'.format(intf)
+	    docker_cmd = asichost.get_docker_cmd(cmd, "database")
```
You could call sonic-db-cli from base image, no need to get into the database docker and call redis-cli from there. Is there a special consideration you did so?

prgeor replied:
@yxieca no special consideration, its just that I found asichost.get_docker_cmd() more convenient to use here. do you see any issue?

yxieca replied:
sonic-db-cli is more generic and less likely to change across versions. But this is not a blocker, we can deal with that when we cross that bridge.
